### PR TITLE
Skip secure settings dependencies during desired nodes settings validation

### LINKED ISSUE #96127: Desired nodes node settings validation fails if a setting depends on a secure setting
During desired nodes updates the system validates that the provided node settings are valid in order to avoid boot looping issues. If one of the node settings depends on a secure setting, the validation fails even when the secure setting is present and correct.

### PR DESCRIPTION
**Pull request by fcofdez**
Closes https://github.com/elastic/elasticsearch/issues/96127

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/settings/Settings.java [hunk 1]
```diff
@@ -133,8 +133,7 @@ private Settings(Map<String, Object> settings, SecureSettings secureSettings) {
     /**
      * Retrieve the secure settings in these settings.
      */
-    SecureSettings getSecureSettings() {
-        // pkg private so it can only be accessed by local subclasses of SecureSetting
+    public SecureSettings getSecureSettings() {
```
Comment (fcofdez) [inline server/src/main/java/org/elasticsearch/common/settings/Settings.java:136]: I'm not super happy about this, but I couldn't find another way to provide the secure settings during the desired nodes validation. I'm happy to hear your feedback.

File: server/src/main/java/org/elasticsearch/cluster/desirednodes/DesiredNodesSettingsValidator.java [hunk 2]
```diff
@@ -104,7 +106,16 @@ private void validate(DesiredNode node) {
             settings = updatedSettings.build();
         }
 
-        clusterSettings.validate(settings, true);
+        final Settings settingsToValidate;
+        if (nodeSettings.getSecureSettings() != null) {
```
Comment (fcofdez) [inline server/src/main/java/org/elasticsearch/cluster/desirednodes/DesiredNodesSettingsValidator.java:110]: I think `nodeSettings.getSecureSettings()` should be always != null?

**Comment by fcofdez** [comment]
Looks like this approach doesn't work 100% of the time, I'm moving the PR back to draft.

**Review by henningandersen — COMMENTED** [review]
I am in doubt about this. I wonder if we should disable the setting validation for now and then work on a solution that can handle secure settings too?

File: server/src/main/java/org/elasticsearch/cluster/desirednodes/DesiredNodesSettingsValidator.java [hunk 3]
```diff
@@ -98,13 +99,27 @@ private void validate(DesiredNode node) {
...
-            final Settings.Builder updatedSettings = Settings.builder().put(settings);
-            updatedSettings.remove(NODE_PROCESSORS_SETTING.getKey());
-            settings = updatedSettings.build();
+            ).get(nodeSettings);
+            settingsToValidate.remove(NODE_PROCESSORS_SETTING.getKey());
         }
 
-        clusterSettings.validate(settings, true);
+        for (String settingKey : nodeSettings.keySet()) {
+            var rawSetting = clusterSettings.getRaw(settingKey);
```
[the remark this answers is not in the document]
Comment (fcofdez) [reply server/src/main/java/org/elasticsearch/cluster/desirednodes/DesiredNodesSettingsValidator.java:107]: This is to cover the case of affix settings; for affix settings the `#get` method returns the materialized setting that doesn't include its dependencies.

File: server/src/main/java/org/elasticsearch/cluster/desirednodes/DesiredNodesSettingsValidator.java [hunk 4]
```diff
@@ -98,13 +99,27 @@ private void validate(DesiredNode node) {
...
+            if (rawSetting == null) {
+                continue;
+            }
+            for (Setting.SettingDependency settingsDependency : rawSetting.getSettingsDependencies(settingKey)) {
+                var dependency = settingsDependency.getSetting();
+                if (dependency.isSecure()) {
+                    // we cannot access to the secure settings after the node has been initialized,
+                    // therefore we cannot validate that the dependency is present at this stage
+                    settingsToValidate.remove(settingKey);
```
[the remark this answers is not in the document]
Comment (fcofdez) [reply server/src/main/java/org/elasticsearch/cluster/desirednodes/DesiredNodesSettingsValidator.java:116]: > Will this not risk moving the problem to another place, i.e., we now remove a setting that some other setting could depend upon?

Yes, that's a possibility.

> Also, we would no longer validate this specific setting, including the "format" of it.

An alternative is to add a settings validation method that doesn't try to resolve the dependencies for secure settings, wdyt?

**Comment by fcofdez** [comment]
> I am in doubt about this. I wonder if we should disable the setting validation for now and then work on a solution that can handle secure settings too?

Maybe we could add a settings validation method that doesn't resolve secure settings instead of removing the setting that depends on them?

**Comment by henningandersen** [comment]
> > I am in doubt about this. I wonder if we should disable the setting validation for now and then work on a solution that can handle secure settings too?
> 
> Maybe we could add a settings validation method that doesn't resolve secure settings instead of removing the setting that depends on them?

I think that could work, though I could imagine a case, where the default value of a setting depends upon the existence of a secure setting. Probably not relevant though and perhaps this is not possible even?

**Comment by fcofdez** [comment]
@elasticmachine update branch

**Comment by fcofdez** [comment]
Superseded by https://github.com/elastic/elasticsearch/issues/96127

**Pull request closed without merging**